### ISSUE TITLE
Css Parser enhancements to CssNode

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#145]
The proposal here (copied from #38) is to have a single, unified set of objects in the DOM and parse tree. The methods used to traverse the two trees are slightly different.

The parse tree preserves the original document, i.e. an inorder traversal of leaf nodes (tokens) produces the source text. In this example the start of the tree looks something like this (I have omitted the whitespace tokens, but they would be in the tree as as well.

```
CssStyleSheet
+ CssStyleRule
  + Selector
    + SimpleSelector
      + Token (identifier): div
    + Token (comment): /* element*/
    + Combinator:
       + Token: >
    + SimpleSelector:
       + Token :foo
```

We just need two different methods to retrieve the source, one that skips comments, and one that preserves them. No reordering is then necessary, and the parse tree has a simple structure with minimum memory overhead - just the `_child` and `_next` pointers.

So yes, it is true that tokens are flat - concatenating them in an inorder traversal of the tree produces the source document. But they are also contained within the parsed tree structure.

Note that this discussion pertains only to the parsed tree, i.e. the `_child` and `_next` pointers. The OM tree is unaffected.

BTW in this scenario Tokens only need one piece of data: the text. The `TextPosition` can be computed as needed (unless this is in the hot path of some usage; I don't need it personally.) In addition `CssTokenType` can be eliminated altogether by just having subclasses of CssToken. This further reduces the memory footprint.

To summarize, here is the set of proposed changes:
- `CssNode` is the ultimate base class of all OM entities and tokens.
- Each token type is a separate class, e.g. `UrlToken`, `ColorToken` (this is basically a memory optimization.)
- The parse tree is composed of `_child` and `_next` pointers of `CssNode`.
- Tokens appear at the correct position in the parse tree.
- An inorder traversal of the parse tree (or any node) produces the stylesheet (or the text of the node.) Comments may be skipped during traversal if desired.

**Comment by FlorianRappl** [comment]
Will work on this later this week / next week.
